packages/provider/src/wallet/MultiChainController.ts

```diff
@@ -0,0 +1,247 @@
+import {
+  BaseControllerV2 as BaseController,
+  RestrictedControllerMessenger,
+} from '@metamask/controllers';
+import { assert, Maybe, Option, Snap, SnapId } from '@metamask/snap-utils';
+import {
+  AccountID,
+  accountIdRe,
+  ChainId,
+  chainIdRe,
+  ConnectArgumentsFull,
+  Namespace,
+  NamespaceId,
+  RequestArguments,
+  Session,
+} from '../client';
+import { findMatchingKeyringSnaps } from './matching';
+
+const controllerName = 'MultiChainController';
+
+const defaultState: MultiChainControllerState = {
+  sessions: {},
+};
+
+type MultiChainControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  any,
+  any,
+  any,
+  any
+>;
+
+type RequestedNamespace = {
+  methods: string[];
+  events: string[];
+  chains: ChainId[];
+};
+
+type SessionData = {
+  origin: string;
+  requestedNamespaces: Record<NamespaceId, RequestedNamespace>;
+  providedNamespaces: Record<NamespaceId, Namespace>;
+  handlingSnaps: Record<NamespaceId, SnapId>;
+};
+
+type MultiChainControllerState = {
+  sessions: { [origin: string]: SessionData };
+};
+
+type Notify = (
+  origin: string,
+  data: { method: string; params?: Record<string, unknown> },
+) => Promise<void>;
+
+type MultiChainControllerArgs = {
+  notify: Notify;
+  messenger: MultiChainControllerMessenger;
+};
+
+function parseChainId(chainId: ChainId): {
+  namespace: NamespaceId;
+  reference: string;
+} {
+  const match = chainIdRe.exec(chainId);
+  assert(match !== null);
```

Comment:
This can be moved to an assert function [here](https://github.com/MetaMask/snaps-skunkworks/blob/184b8050d685aaab3bd65f0f9923df3552c68be4/packages/utils/src/namespace.ts), e.g.,

```ts
import { assert } from 'superstruct';

export function assertIsChainId(value: unknown): asserts value is ChainId {
  assert(value, ChainIdStruct);
}
```